src/librustc_resolve/diagnostics.rs

```diff
@@ -643,22 +643,24 @@ impl<'a> Resolver<'a> {
         let not_local_module = crate_name.name != kw::Crate;
         let mut worklist =
             vec![(start_module, Vec::<ast::PathSegment>::new(), true, not_local_module)];
+        let mut worklist_via_import = vec![];
 
-        while let Some((in_module, path_segments, accessible, in_module_is_extern)) = worklist.pop()
+        while let Some((in_module, path_segments, accessible, in_module_is_extern)) =
+            match worklist.pop() {
+                None => worklist_via_import.pop(),
+                Some(x) => Some(x),
+            }
```

Comment:
I'm not sure what is the logic here - prioritizing non-imports over imports? I'm not sure this is useful, e.g. imports and non-imports from other crates are not differentiated here (for technical reasons) and it seems working ok.

Reply:
It's also not clear why variants get special treatment.

Reply:
These two are to prevent side-effects:

1. There is an assumption elsewhere that paths of variants are in the enum's declaration and not imported. With this assumption, the variant component is chopped and the rest of the path is assumed to be the enum's own path. For errors where a variant is used as the type instead of the enum, this causes funny looking invalid suggestions, i.e `foo` instead of `foo::MyEnum`.
2. Non-imports are prioritized, otherwise weird re-exports paths are picked instead, like `use io::sys::ext::net::sys_common::util::thread::panic::collections::HashMap;`, instead of just `use std::collections::HashMap;`.